src/ttt/ttt_game.rs

```diff
@@ -326,4 +340,64 @@ impl IGame for TttGame {
     fn get_repetition_limit() -> Option<u32> {
         None
     }
+
+    fn produce_transformed_data_entries(entry: DataEntry<Self>) -> Vec<DataEntry<Self>> {
+        let transform = |e: &DataEntry<Self>, transform_sq: &dyn Fn(usize) -> usize| {
+            let (board_x, board_o) = [e.pos.board_x, e.pos.board_o]
+                .iter()
+                .map(|b| {
+                    let mut bt = TttBitboard::new();
+                    for idx in 0..TttGame::BOARD_SIZE * TttGame::BOARD_SIZE {
+                        bt.set(transform_sq(idx), b.get(idx));
+                    }
+                    bt
+                })
+                .collect_tuple()
+                .unwrap();
+            let pos = TttPosition::from_bitboards(board_x, board_o, e.pos.get_turn());
+
+            let probs = e
+                .probs
+                .iter()
+                .map(|(m, p)| (TttMove::from_idx(transform_sq(m.to_idx())), *p))
+                .collect_vec();
+
+            let winner = e.winner;
+            DataEntry { pos, probs, winner }
+        };
+
+        let rows_mirror = |e: &DataEntry<Self>| {
+            transform(e, &|idx| {
+                let r = idx / TttGame::BOARD_SIZE;
+                let c = idx % TttGame::BOARD_SIZE;
+                let rt = TttGame::BOARD_SIZE - 1 - r;
+                let ct = c;
+                rt * TttGame::BOARD_SIZE + ct
+            })
+        };
+        let columns_mirror = |e: &DataEntry<Self>| {
+            transform(e, &|idx| {
+                let r = idx / TttGame::BOARD_SIZE;
+                let c = idx % TttGame::BOARD_SIZE;
+                let rt = r;
+                let ct = TttGame::BOARD_SIZE - 1 - c;
+                rt * TttGame::BOARD_SIZE + ct
+            })
+        };
+        let diagonal_mirror = |e: &DataEntry<Self>| {
```

Comment:
i think there are more than 3 possible transforms... 
there is a secondary diagonal, also there are rotations, and more than one action can be performed
i think there are a total of 8 possible transforms (including the original entry)? i am imagining a transparent 2d board that you can physically flip and rotate
would we want to add all of them to the training data?

Reply:
original
rows
columns
diagoanl
rows + columns = rotation 180
rows + diagoanl = rotate 90
columns + diagoanl = rotate 90 (other direction)
rows + columns + diagoanl = other diagoanl

8 total. the following line:
entries.extend(entries.iter().map({transform}).collect_vec())
extends the vector by transforming each existing one, so doubling the number of elements in the vector, we take all the combinations not just 3

Reply:
beautiful